crates/ruma-client-api/src/r0/uiaa.rs

```diff
@@ -152,6 +152,52 @@ impl<'a> AuthData<'a> {
 }
 
 impl IncomingAuthData {
+    /// Creates a new `IncomingAuthData` with the given `auth_type` string, session and data.
+    ///
+    /// Prefer to use the public variants of `IncomingAuthData` where possible; this constructor is
+    /// meant be used for unsupported message types only and does not allow setting arbitrary
+    /// data for supported ones. This method can't construct the `RegistrationToken` variant.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the `auth_type` is known and serialization of `data` to the
+    /// corresponding `IncomingAuthData` variant fails.
+    pub fn new(
+        auth_type: &str,
+        session: Option<String>,
+        data: JsonObject,
+    ) -> serde_json::Result<Self> {
+        fn deserialize_variant<T: DeserializeOwned>(
+            session: Option<String>,
+            mut obj: JsonObject,
+        ) -> serde_json::Result<T> {
+            if let Some(session) = session {
+                obj.insert("session".into(), session.into());
+            }
```

Comment:
Oh, this just made me realize that we should be doing the same thing in `MessageType::new` (where the `body` parameter has a very similar role as `session` here). 👍🏼